forc-pkg/src/pkg.rs

```diff
@@ -373,6 +373,75 @@ pub struct BuildOpts {
     pub tests: bool,
     /// List of constants to inject for each package.
     pub const_inject_map: ConstInjectionMap,
+    /// The set of options to filter by member project kind.
+    pub build_filter: BuildFilter,
+}
+
+/// The set of options to filter type of projects to build in a workspace.
+#[derive(Default)]
+pub struct BuildFilter {
+    pub build_contracts: bool,
+    pub build_scripts: bool,
+    pub build_predicates: bool,
+    pub build_libraries: bool,
+}
+
+impl BuildFilter {
+    /// Returns a new `BuildFilter` that only builds scripts.
+    pub fn only_scripts() -> Self {
+        Self {
+            build_contracts: false,
+            build_scripts: true,
+            build_predicates: false,
+            build_libraries: false,
+        }
+    }
+
+    /// Returns a new `BuildFilter` that only builds contracts.
+    pub fn only_contracts() -> Self {
+        Self {
+            build_contracts: true,
+            build_scripts: false,
+            build_predicates: false,
+            build_libraries: false,
+        }
+    }
+
+    /// Returns a new `BuildFilter` that builds all types of projects.
+    pub fn all() -> Self {
+        Self {
+            build_contracts: true,
+            build_scripts: true,
+            build_predicates: true,
+            build_libraries: true,
+        }
+    }
+
+    /// Filter given target of output nodes according to the this `BuildFilter`.
+    pub fn filter_outputs(
+        &self,
+        build_plan: &BuildPlan,
+        outputs: HashSet<NodeIx>,
+    ) -> HashSet<NodeIx> {
+        let graph = build_plan.graph();
+        let manifest_map = build_plan.manifest_map();
+        outputs
+            .into_iter()
+            .filter(|&node_ix| {
+                let pkg = &graph[node_ix];
+                let pkg_manifest = &manifest_map[&pkg.id()];
+                let program_type = pkg_manifest
+                    .program_type()
+                    .expect("cannot retrieve program type for filtering");
```

Comment:
We might want to return an error for this case rather than `panic!`? It's not unlikely we might hit this as currently checking the `program_type()` requires doing I/O and parsing the Sway file, which might have changed since the manifest was originally loaded. Ideally we'll one day put the program kind in the manifest instead (I think this might have been mentioned in https://github.com/FuelLabs/sway-rfcs/pull/8).